### PR TITLE
closes #950: refactors extensions docs + adds ExtensionGroup documentation

### LINKED ISSUE #950: Re-structure extensions documentation in integrators-guide
In a recent issue I found out the use of `ExtensionGroups`, and saw it's not documented.
I wanted to add it, but then though it could be a chance to improve the overall section to accomodate the new docs.

Right now, to me how it's explained is a bit confusing.
At the beginning of the section we say:
```
To create an extension two things are required:

1. Create a class extending one of the extension classes from above
2. Register your class using the JavaExtensionRegistry class
```

But, then we jump into the AST stating knowing about it is also a pre-requisit.
After that, we explain at the same section depth:

* Each extension type and for each how to register directly using `asciidoctor.javaExtensionRegistry()`.
* How register with SPI automatically

This is the tree as seen in Intellj plugin (excuse the bug showing the cols and note as sections):
image

I think it would be easier if we beging saying there are "three steps":
* Briefly understand how Asciidoctor converts documents -> we can talk about AST + how extension plug into it without much detail since extensions are explained later on it's own section. We should not make this cumbersome or overly complex not to discourage people, I can imagine that hearing AST right at the start can be dauning for some.
* How to create an AsciidoctorJ extension -> here we aggregate the extension types, and remove the duplicated examplations about `asciidoctor.javaExtensionRegistry()` from each.
* How to register your extension -> here, in 3 subsections (level 4 in depth) we explain:
  * Manual registering (so we remove duplciation in previous section)
  * SPI
  * ExtensionGroups

I tend to have this rule in which if an introduction mentions a list of key elements, then each point/number should map to a sub-section. wdyt?

Related to this I have some questions:
* Should we recomend `ExtensionGroups` above the other methods? It seems to me this is simpler and more flexible since it allows un-registering, and also does not leak the java implementation. Are there any pros-cons of using manual vs the groups? _For SPI there's a clear advantage of automatic detection_
* Is there a set of JVM languages we can mention with confidence other than Java and Groovy? This is OSS so I don't think we need to fully commit or talk about "support", but i'd be nice if we can mantion some that we feel confident about.

### PR DESCRIPTION
**Pull request by abelsromero**
Thank you for opening a pull request and contributing to AsciidoctorJ!

Please take a bit of time giving some details about your pull request:

## Kind of change

- [ ] Bug fix
- [ ] New non-breaking feature
- [ ] New breaking feature
- [x] Documentation update
- [ ] Build improvement

## Description

What is the goal of this pull request?
Apply changes in #950

How does it achieve that?

Are there any alternative ways to implement this?
Sure, there's a degree of subjetivity in docs. But avoid this, I tried to keep changes simple.

Are there any implications of this pull request? Anything a user must know?
No

## Issue

If this PR fixes an open issue, please add a line of the form:

Fixes #Issue 

## Release notes

Please add a corresponding entry to the file CHANGELOG.adoc[](url)

### CODE REVIEW

**Comment by abelsromero** [comment]
This is WIP, I will review tomorrow.

**Comment by abelsromero** [comment]
Also I am not sure I am modifying the apropiate files, since they are duplicated under _docs_ and _asciidoctorj-documentation_ project. @robertpanzer how should I do?

**Comment by robertpanzer** [comment]
The idea is that you edit the files under asciidoctorj-documentation and run the build. 
This generates the files und docs resolving the include macros as GitHub doesn't do that.

**Comment by abelsromero** [comment]
@robertpanzer  done! modified the sources in _asciidoctorj-documentation_ project and run the `processAsciidocFiles` task. If that's the process, I'd like to make another PR explainning this in the README.

File: asciidoctorj-documentation/src/main/asciidoc/integrator-guide.adoc [hunk 1]
```diff
@@ -499,7 +498,16 @@ One of the major improvements to Asciidoctor recently is the extensions API.
...
+To create an extension we only need to:
+
+. Have a brief understanding of the Asciidoctorj conversion process and internal representation of the document.
+In short, extensions allow you to tap into the conversion process to make modifications to enhance the generated content.
+. Write an extension class.
+This is as simple as creating a sub-class from one of the extension points listed below.
+. Register your class to make Asciidoctorj converter aware of it.
+There several options depending on the use case.
+
+Asciidoctor provides eighth types of extension points.
```
Comment (robertpanzer) [inline asciidoctorj-documentation/src/main/asciidoc/integrator-guide.adoc:510]: s/eighth/eight/

File: asciidoctorj-documentation/src/main/asciidoc/integrator-guide.adoc [hunk 2]
```diff
@@ -988,30 +957,24 @@ include::src/test/java/org/asciidoctor/integrationguide/extension/CopyrightFoote
 ----
 <1> All Preprocessors must extend the class `org.asciidoctor.extension.Postprocessor` and implement the method `process()`.
```
Comment (robertpanzer) [inline asciidoctorj-documentation/src/main/asciidoc/integrator-guide.adoc:958]: Not related to your PR, but could you fix this? 😁 
All ~Preprocessors~ Postprocessors must extend the class `org.asciidoctor.extension.Postprocessor`

**Review by robertpanzer — COMMENTED** [review]
Thank you, this is awesome!
I just added some comments for a few typos.
If you could fix these I'll happily merge this!

File: asciidoctorj-documentation/src/main/asciidoc/integrator-guide.adoc (hunk 2 above)
Comment (abelsromero) [reply asciidoctorj-documentation/src/main/asciidoc/integrator-guide.adoc:958]: I can tell from this one, that you did go through all the text in defail

File: asciidoctorj-documentation/src/main/asciidoc/integrator-guide.adoc [hunk 3]
```diff
@@ -827,7 +818,7 @@ class CiteInlineMacroProcessor extends InlineMacroProcessor {
 With the SHORT format, the attributes are not parsed, and the 'target' that is passed to your macro is the value between the brackets ("brown79").
 
 [[blockprocessor]]
-=== Block Processors
+===== Block Processors
```
[the remark this answers is not in the document]
Comment (abelsromero) [reply asciidoctorj-documentation/src/main/asciidoc/integrator-guide.adoc]: corrected, thanks

File: asciidoctorj-documentation/src/main/asciidoc/integrator-guide.adoc [hunk 4]
```diff
@@ -548,38 +556,39 @@ a|Processes an arbitary block based on it's style such as
...
-* Block macro, inline macro, and block processors are called in the order that they appear in the document.
+Before starting to write your first extension, some basic understanding of how Asciidoctor treats the document is helpful.
+As any language processing tool, the process can be roughly split into three steps:
 
-* Prepocessors, Treeprocessors, Postprocessors and DocinfoProcessors are called in an arbitrary and changeable order.
---
+. Parsing: the raw sources content is read and analyzed to generate the internal representation, the AST (_abstract syntax tree_).
+. Processing: the AST is processed. For example to detect possible errors, add automatically generated content (toc), etc.
+. Output generation: once the final AST is set, it's again processed to generate the desired output.
+For example, a sub-section of the AST representing a title with a paragraf will be converted into it's correspondent HTML or PDF output.
```
[the remark this answers is not in the document]
Comment (abelsromero) [reply asciidoctorj-documentation/src/main/asciidoc/integrator-guide.adoc]: fixed :+1:

File: asciidoctorj-documentation/src/main/asciidoc/integrator-guide.adoc [hunk 5]
```diff
@@ -548,38 +556,39 @@ a|Processes an arbitary block based on it's style such as
...
 
-To create an extension two things are required:
+NOTE: Some liberty is taken to make the process easier to understand.
+In truth, Asciidoctor has implementation details that divert from the 3 steps above.
```
[the remark this answers is not in the document]
Comment (abelsromero) [reply asciidoctorj-documentation/src/main/asciidoc/integrator-guide.adoc]: > native speaker here

probably, such subtleties scape me :man_shrugging:

File: asciidoctorj-documentation/src/main/asciidoc/integrator-guide.adoc (hunk 1 above)
Comment (abelsromero) [reply asciidoctorj-documentation/src/main/asciidoc/integrator-guide.adoc]: :man_facepalming: thanks

**Comment by abelsromero** [comment]
All corrections applied, I squashed the change

**Comment by robertpanzer** [comment]
Awesome, thank you!

**Pull request merged**